src/de.hpi.swa.trufflesqueak/src/de/hpi/swa/trufflesqueak/model/EphemeronObject.java

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright (c) 2017-2025 Software Architecture Group, Hasso Plattner Institute
+ * Copyright (c) 2021-2025 Oracle and/or its affiliates
+ *
+ * Licensed under the MIT License.
+ */
+package de.hpi.swa.trufflesqueak.model;
+
+import java.lang.ref.ReferenceQueue;
+import java.lang.ref.WeakReference;
+
+import com.oracle.truffle.api.CompilerAsserts;
+import com.oracle.truffle.api.CompilerDirectives;
+import com.oracle.truffle.api.CompilerDirectives.TruffleBoundary;
+
+import de.hpi.swa.trufflesqueak.image.SqueakImageChunk;
+import de.hpi.swa.trufflesqueak.image.SqueakImageContext;
+import de.hpi.swa.trufflesqueak.image.SqueakImageWriter;
+import de.hpi.swa.trufflesqueak.model.layout.ObjectLayouts;
+import de.hpi.swa.trufflesqueak.util.ArrayUtils;
+import de.hpi.swa.trufflesqueak.util.ObjectGraphUtils.ObjectTracer;
+
+public final class EphemeronObject extends AbstractSqueakObjectWithClassAndHash {
+
+    private final SqueakImageContext image;
+
+    @CompilerDirectives.CompilationFinal private Object key;
+    @CompilerDirectives.CompilationFinal private Object value;
```

Comment:
Does it really make sense to mark `key` and `value` as `CompilationFinal`?

Reply:
I don't know the full implications of `CompilationFinal` -- I just copied some of the implementation of `ClassObject` and applied it here. What do you think would be best?

Reply:
I'd remove it, unless you have a benchmark that shows this is actually useful. What this annotation does is mark a field as a constant at runtime iff the Ephemeron object itself is a constant. I'm not sure this is what we want for `key` and `value`. They are potentially mutable and the additional `transferToInterpreterAndInvalidate` are probably causing more harm (transfer to interpreter and invalidate code) than they are useful.

Reply:
I agree and have already changed it.